modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoDataLookupFactory.java

```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.ingest.geoip;
+
+import java.util.Set;
+
+@FunctionalInterface
+interface GeoDataLookupFactory<PROVIDER extends GeoDataLookup> {
+    PROVIDER create(Set<Database.Property> properties);
+
+    static GeoDataLookupFactory<?> get(final Database database) {
```

Comment:
`Database` still owns the mapping of which properties are available/default for each of these lookups. We could likely benefit from either consolidating them or making `Database` implement `GeoDataLookupFactory` instead of requiring this separate 1:1 lookup.